### ISSUE TITLE
session.clear not working across apps

### ISSUE THREAD

**Issue by eiko** [jeremyevans/roda#162]
I have a main App and an Auth app for handling authentication:

```ruby
class App < Roda
  plugin :sessions, secret: 'mysecret'
  
  route do |r|
    r.on 'auth' do
      r.run Auth
    end
    # app routes
  end
end
```

```ruby
class Auth < Roda
  plugin :sessions, secret: 'mysecret'

  route do |r|
    r.on 'logout' do
      session.clear
      r.redirect '/'
    end
  end
end
```

When I set session data in Auth, I can access it in App. And when I call `session.clear` in App, it properly clears. But for some reason, calling `session.clear` in Auth does not clear data after redirecting back to App. Am I doing something wrong?

**Comment by jeremyevans** [comment]
You probably don't need the plugin :sessions call in `Auth`. Does it work correctly if you remove that?  Is it possible for you to submit a failing spec for the behavior you are seeing?

**Comment by eiko** [comment]
Removing `plugin :sessions` gives me:

```
Roda::RodaError: You're missing a session handler, try using the sessions plugin.
```

I can make a spec! Would you prefer it to be a pull request with the spec in the same folder as the others?

**Comment by jeremyevans** [comment]
Sure, that would be great.  I'm not sure this is a bug in Roda, hopefully I can look at the spec and determine if this is a bug or if is something that cannot be supported by the implementation and needs to be documented.

Note that for situations like this with more than one application, you may want to use `roda/session_middleware` instead of the sessions plugin.

For the particular error that occurs when not using the session plugin in Auth, you could also try the following in App:

```ruby
    r.on 'auth' do
      session # force loading of session
      r.run Auth
    end
```

**Comment by eiko** [comment]
As it turns out, the problem isn't with multiple apps. Working on a sample app gave me insight on the actual problem, which is this:

```ruby
class App < Roda
  plugin :sessions, secret: 'mysecret'
  
  route do |r|
    r.on 'set' do
      session['foo'] = 'bar'
      'data set'
    end

    r.on 'clear' do
      clear_session
      'this works'
    end

    r.on 'nested/clear' do
      clear_session
      'this does not work!'
    end
  end
end
```

The cookie can be cleared from root and its direct children, but not from any indirect children. I believe this has something to do with the `path` aspect of cookies, but I'm still looking through roda's source for a bit of code that confirms it.

**Comment by jeremyevans** [comment]
Odd, the default cookie options for the sessions cookie include `:path=>'/'`.

**Comment by eiko** [comment]
I found the root of the problem, which is:

```ruby
# lib/roda/plugins/sessions.rb#236
Rack::Utils.delete_cookie_header!(headers, opts[:key])

#=> {"Content-Type"=>"text/html", "Content-Length"=>"14", "Set-Cookie"=>"roda.session=; max-age=0; expires=Thu, 01 Jan 1970 00:00:00 -0000"}
```

The Rack method omits `path`, so the browser defaults to the path of the current page, which is why removing cookies only works on pages where the root path is `/`. I added a line of code to the plugin that appends `path=/` to the header, and that fixed the issue.

I'm almost off the clock, but I'll create a proper spec and pull request some time this weekend.

**Comment by jeremyevans** [comment]
Thank you for the analysis.  I agree that this is a bug and it sounds like passing `opts[:cookie_options]` as the third argument to `delete_cookie_header!` should fix it.  This way it will also fix cases where `:domain` is used.

**Comment by jeremyevans** [comment]
Closed by #163.
